### DISCUSSION TITLE
Wrote a utility script to help process large text files for longer narration.

### DISCUSSION THREAD

**Discussion by TheMechanicX32** [coqui-ai/TTS#3801] [General]
Good morning all!

I wrote this helper function to help me generate audio from long text files. Here is the script:

```
import re

def split_text_into_chunks(text, desired_chunk_length=200, max_chunk_length=250):
   """Split text into chunks of a desired length trying to keep sentences intact."""
   # Normalize text, remove redundant whitespace and convert non-ascii quotes to ascii
   text = re.sub(r'\n\n+', '\n', text)
   text = re.sub(r'\s+', ' ', text)
   text = re.sub(r'[“”]', '"', text)

   chunks = []
   is_within_quotes = False
   current_chunk = ""
   sentence_end_positions = []
   current_position = -1
   text_end_position = len(text) - 1

   def move_position(delta):
      nonlocal current_position, is_within_quotes, current_chunk
      is_negative = delta < 0
      for _ in range(abs(delta)):
         if is_negative:
               current_position -= 1
               current_chunk = current_chunk[:-1]
         else:
               current_position += 1
               current_chunk += text[current_position]
         if text[current_position] == '"':
               is_within_quotes = not is_within_quotes
      return text[current_position]

   def commit_chunk():
      nonlocal chunks, current_chunk, sentence_end_positions
      chunks.append(current_chunk.strip())
      current_chunk = ""
      sentence_end_positions = []

   while current_position < text_end_position:
      char = move_position(1)
      # Record sentence end positions
      if char in '.!?':
         sentence_end_positions.append(current_position)
      # Commit chunk if it's the right size and at a sentence end
      if len(current_chunk) >= desired_chunk_length and current_position in sentence_end_positions:
         commit_chunk()
      # If the chunk is too long, find the last sentence end and commit up to there
      elif len(current_chunk) >= max_chunk_length:
         last_sentence_end = max([pos for pos in sentence_end_positions if pos < current_position], default=-1)
         if last_sentence_end != -1:
               # Move back to the last sentence end
               move_position(last_sentence_end - current_position)
               commit_chunk()
         else:
               # If no sentence end found, commit the whole chunk and start a new one
               commit_chunk()

   # Commit any remaining text as a chunk
   if current_chunk:
      commit_chunk()

   # Clean up, remove lines with only whitespace or punctuation
   chunks = [chunk for chunk in chunks if len(chunk) > 0 and not re.match(r'^[\s\.,;:!?]*$', chunk)]

   for chunk in chunks:
      print(chunk, len(chunk), "\n")
      if len(chunk) > 250:
         print("\nERROR\n")
      
   return chunks
```
   
Usage:

Create a new python script and call it `text.py` (or whatever suits your needs). Drop it in the same directory as the script you use to run XTTS. Inside your run script, import the function from `text.py` like so: `from text import split_text_into_chunks;`

To use the text splitter, pull data from a text file and pipe it into the `split_text_into_chunks()` function like so:

    with open("input_text.txt", "r") as file:
       text = file.read();

    separated_input = split_text_into_chunks(text);

I have to use `torchaudio` to concatenate all my audio files together after the script finishes generating all the audio segments. My final `run.py` script looks like this:

```
import os
import torch
import torchaudio
from TTS.api import TTS
from time    import time
from text    import split_text_into_chunks;

# Get device
device = "cuda" if torch.cuda.is_available() else "cpu"

with open("input_text.txt", "r") as file:
   text = file.read();

separated_input = split_text_into_chunks(text);

# List available 🐸TTS models
print(TTS().list_models())

# Init TTS
tts = TTS("tts_models/multilingual/multi-dataset/xtts_v2").to(device)

audio_clips = [];

# Run TTS and save each audio clip to a file
output_path = 'output'  # Set the path where you want to save the audio clips

start_time = time()

for index, text in enumerate(separated_input):
   # Since this model is multi-lingual voice cloning model, we must set the target speaker_wav and language
   # Generate text to speech list of amplitude values as output
   wav = tts.tts(text=text, speaker_wav="voice.wav", language="en")
   
   # Convert the list to a PyTorch tensor
   wav_tensor = torch.tensor(wav)

   # Check if the tensor is 1D and add a channel dimension if necessary
   if wav_tensor.dim() == 1:
      wav_tensor = wav_tensor.unsqueeze(0)
   
   # Save each tensor as a WAV file
   audio_file_path = os.path.join(output_path, f"{index}.wav")
   torchaudio.save(audio_file_path, wav_tensor, sample_rate=24000)

# Load and concatenate the saved audio clips
all_audio_parts = []
for index in range(len(separated_input)):
   audio_file_path = os.path.join(output_path, f"{index}.wav")
   audio_part, _ = torchaudio.load(audio_file_path)  # Load the audio file
   all_audio_parts.append(audio_part)

# Concatenate all audio parts along the time axis (usually the last dimension)
concatenated_audio = torch.cat(all_audio_parts, dim=-1)

# Save the concatenated audio to a file
final_output_path = os.path.join(output_path, "audio.wav")
torchaudio.save(final_output_path, concatenated_audio, sample_rate=24000)

end_time = time()

print("Total time taken: ", round(end_time-start_time), "seconds")
```

This utility script I wrote splits the text file into sentences. It does it's hardest to avoid splitting in the middle of a sentence, which causes really strange flow in the final product. Makes the audio hard to follow while listening. Generally speaking, the tts model does best with sentences that are not abnormal. Clear, concise, neat grammar is rewarded here. The utility function relies on correct punctuation. Let me know if you have any questions. I don't check github too often, but I'll try to be more active in this thread.

**Comment by nucklearproject** [comment]
Thanks! is very Useful! 😎👌

**Comment by ci21-cgarijo** [comment]
Very useful, and nice code! Exactly what I was looking for. 💯👌

**Comment by nequamDev** [comment]
I was trying to make this for a week I did it but still running into some bug with non-ascii this is really helpful thanks
